include/caffe/datum_DB.hpp

```diff
@@ -0,0 +1,128 @@
+#ifndef CAFFE_DATUMDB_HPP
+#define CAFFE_DATUMDB_HPP
+
+#include <string>
+
+#include "leveldb/db.h"
+#include "lmdb.h"
+
+#include "caffe/common.hpp"
+#include "caffe/datum_DB_factory.hpp"
+#include "caffe/proto/caffe.pb.h"
+
+namespace caffe {
+
+class DatumDB {
+ public:
+  explicit DatumDB(const DatumDBParameter& param)
+    : param_(param),
+      is_opened_(new bool(false)) {}
+  virtual ~DatumDB() {
+    if (is_opened_.unique()) {
+      DatumDBRegistry::RemoveSource(param_.source());
+    }
+  }
+
+  class Generator {
```

Comment:
Yeah the `Generator` is iterator-like abstraction but simpler, it could be out `DatumDB` and being called `DatumGenerator` for clarity.
I explored different options, and the iterator-like need to have an state, which at the end pretty much a `DatumDB` so, instead of creating a new state, a `Generator` just receives a copy of `DatumDB` which holds the needed state.